src/Compilers/Server/VBCSCompilerTests/CompilerServerTests.cs

```diff
@@ -109,14 +109,18 @@ private static void ReferenceNetstandardDllIfCoreClr(List<string> arguments)
 #endif
         }
 
+        private static readonly object CreateFilesLock = new object();
         private static void CreateFiles(TempDirectory currentDirectory, IEnumerable<KeyValuePair<string, string>> files)
         {
             if (files != null)
             {
-                foreach (var pair in files)
+                lock (CreateFilesLock)
                 {
-                    TempFile file = currentDirectory.CreateFile(pair.Key);
-                    file.WriteAllText(pair.Value);
+                    foreach (var pair in files)
+                    {
+                        TempFile file = currentDirectory.CreateFile(pair.Key);
```

Comment:
Since CreateFile is the not thread safe line, could you simply lock around it, and return the pointer to the file, and then do the IO to the file? (Locking inside the loop, instead of around it.)
Locking the WriteAllText seems like the lock would be held for longer than required, if the various tasks would all be writing to different files.
Main cost of locks is waiting for them to free, not actually taking/releasing them, therefore taking them often, but locking for the minimal amount of time often gives the most benefits from having parallel execution.

Reply:
That's theoretically possible, yes. However, this is in a test, not in production code, and that level of optimization isn't really needed, and just complicates maintainability. (Files being written here are also probably <1KB on average)

If we really cared about performance here, a better solution would be to make `TempRoot` use a collection designed for concurrency and performance.